presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java

```diff
@@ -98,26 +106,35 @@ public void add(Block block, int position)
         checkArgument(position >= 0, "position must be >= 0");
 
         if (block.isNull(position)) {
-            containsNullElement = true;
-        }
-        else {
-            int hashPosition = getHashPositionOfElement(block, position);
-            if (blockPositionByHash.get(hashPosition) == EMPTY_SLOT) {
-                addNewElement(hashPosition, block, position);
+            if (!containsNullElement) {
+                containsNullElement = true;
             }
         }
+
+        // null is added to both elementBlock and the blockPositionByHash hash table only once.
+        // Adding it to elementBlock makes getBlock() being able to be called multiple times.
+        // Adding it to hash table avoids changes in rehash()
+        int hashPosition = getHashPositionOfElement(block, position);
+        if (blockPositionByHash.get(hashPosition) == EMPTY_SLOT) {
+            addNewElement(hashPosition, block, position);
+        }
     }
 
     public int size()
     {
-        return elementBlock.getPositionCount() + (containsNullElement ? 1 : 0);
+        return elementBlock.getPositionCount() - elementBlockStartOffset;
     }
 
     public int positionOf(Block block, int position)
     {
         return blockPositionByHash.get(getHashPositionOfElement(block, position));
     }
 
+    public Block getBlock()
```

Comment:
I thought it would be better to encapsulate the logic in TypedSet in this method. But if we want to avoid changes to TypedSet as much as possible then I'll remove it. Otherwise I think it makes sense for TypedSet to return the built Block, because caller has passed in the blockBuilder to TypedSet and it's TypedSet's responsibility to return a Block from this blockBuilder.

Reply:
That's a good point. The name `getBlock` is a bit confusing though, since it makes me think it returns the underlying data container block (instead of the region block).

Let's keep the current solution (i.e. do the slice outside `TypedSet`) for now.